File: src/test/java/info/dreamcoder/jdbc/DatabaseTest.java
package info.dreamcoder.jdbc;

import org.assertj.db.type.Request;
import org.assertj.db.type.Source;
import org.junit.jupiter.api.*;

import org.mockito.internal.util.reflection.FieldSetter;

import java.sql.Connection;
import java.sql.SQLException;

import static org.assertj.core.api.Assertions.*;
import static org.assertj.db.api.Assertions.*;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@DisplayName("数据库操作相关测试")
class DatabaseTest {
    private Database database;

    @BeforeEach
    public void createDatabase() {
        try {
            database = new Database(
                    System.getenv("MYSQL_URL"),
                    System.getenv("MYSQL_USERNAME"),
                    System.getenv("MYSQL_PASSWORD")
            );
            database.execute("drop table if exists test_table;");
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    @AfterEach
    public void closeDatabase() {
        database.close();
    }

    @BeforeEach
    public void cleanDatabase() throws SQLException {
        database.execute("drop table if exists test_table;");
    }

    @Test
    @DisplayName("可以正常的执行show table语句")
    void shouldExecuteShowTableSql() {
        assertThatNoException().isThrownBy(() -> database.execute("show tables"));
        assertThat(database.execute("show tables")).isTrue();
    }

    @Test
    @DisplayName("sql错误的时候,得到false的结果")
    void shouldGetFalseWheSqlHasError() {
        assertThat(database.execute("show tables 123")).isFalse();
    }


    @Test
    @DisplayName("能正确创建表格")
    void shouldCreateTable() throws SQLException {
        assertThat(database.execute("create table test_table(id integer)")).isTrue();

        Source source = new Source(
                System.getenv("MYSQL_URL"),
                System.getenv("MYSQL_USERNAME"),
                System.getenv("MYSQL_PASSWORD")
        );

        Request request = new Request(source, "show tables like '%test_table%'");

        assertThat(request).row(0)
                .value().isEqualTo("test_table");
    }

    @Test
    @DisplayName("可以正常的删除表格")
    void shouldDropTable() {
        assertThat(database.execute("create table test_table(id integer)")).isTrue();
    }

    @Test
    @DisplayName("能获取数据库连接状态")
    void shouldGetTrueWhenConnectionIsSuccess() {
        assertThat(database.isValid()).isTrue();
    }

    @Test
    @DisplayName("数据库连接有问题的时候返回false")
    void shouldGetFalseWhenConnectionIsError() throws NoSuchFieldException, SQLException {
        Connection connection = mock(Connection.class);
        FieldSetter.setField(database, database.getClass().getDeclaredField("dbConnection"), connection);
        when(connection.isValid(1000)).thenThrow(SQLException.class);
        assertThat(database.isValid()).isFalse();
    }
}

File: migration-core/src/test/kotlin/jdbc/DatabaseTest.kt
package jdbc

import config.DbType
import org.amshove.kluent.invoking
import org.amshove.kluent.shouldBeFalse
import org.amshove.kluent.shouldBeTrue
import org.amshove.kluent.shouldNotThrow
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import java.sql.SQLException as SQLException

class DatabaseTest {
    val database = Database(DbType.Sqlite)

    @BeforeEach
    fun initTable() {
        database.execute("drop table if exists test_table")
    }

    @Test
    @DisplayName("函数isValid, 数据库能1秒连接上的时候返回 true")
    fun isValid() {
        invoking {
            val db = Database(DbType.Sqlite)
            db.close()
        } shouldNotThrow SQLException::class
    }

    @Test
    @DisplayName("函数close能正确关闭数据库")
    fun close() {
        invoking {
            val db = Database(DbType.Sqlite)
            db.close()
        } shouldNotThrow SQLException::class
    }

    @Test
    @DisplayName("函数execute正确的执行创建和删除表的sql语句")
    fun execute() {
        database.execute("create table test_table_1(name int)")
        database.tableExists("test_table_1").shouldBeTrue()
        database.execute("drop table test_table_1")
        database.tableExists("test_table_1").shouldBeFalse()
    }

    @Test
    @DisplayName("函数query")
    fun query() {
        database.execute("create table test_table_1(name int)")
        database.tableExists("test_table_1").shouldBeTrue()
    }
}